### PR TITLE
Inotify exit watch

### PR DESCRIPTION
**Pull request by mrunalp**
We add an inotify based container exit watch that can asynchronously update the status of the containers that have exited so we don't have to call runtime state on containers during container list which gets called frequently by the kubelet.

Replaces #747

### CODE REVIEW

File: oci/oci.go [hunk 1]
```diff
@@ -579,7 +579,7 @@ func (r *Runtime) UpdateStatus(c *Container) error {
 	}
 
 	if c.state.Status == ContainerStateStopped {
-		exitFilePath := filepath.Join(c.bundlePath, "exit")
+		exitFilePath := filepath.Join("/var/run/crio/exits", c.id)
```
Comment (rhatdan) [inline oci/oci.go]: Can you make this a constant somewhere so that it can be shared between functions?

File: server/server.go [hunk 2]
```diff
@@ -32,6 +34,7 @@ import (
 const (
 	runtimeAPIVersion = "v1alpha1"
 	shutdownFile      = "/var/lib/crio/crio.shutdown"
+	containerExitsDir = "/var/run/crio/exits"
```
Comment (rhatdan) [inline server/server.go]: Perhaps make this global and have oci/oci.go share it or vice versa.

**Comment by rhatdan** [comment]
Should this code eventually be moved to libpod?  Should conman write status to /var/run/libpod/ o r /var/run/containers?  ( would prefer the second.)  conman should really not get tied to crio, perhaps the directory should be passed in via command line?  Will crio care about exit status of containers launched by kpod? or libpod?

**Comment by rhatdan** [comment]
Other then my comments which can be fixed after this is merged, This patch LGTM.

**Comment by mrunalp** [comment]
Pushed update to address comments.

**Comment by rhatdan** [comment]
LGTM

File: server/server.go [hunk 3]
```diff
@@ -286,3 +295,43 @@ func (s *Server) CreateMetricsEndpoint() (*http.ServeMux, error) {
...
+			case event := <-watcher.Events:
+				logrus.Debugf("event: %v", event)
+				if event.Op&fsnotify.Create == fsnotify.Create {
+					containerID := filepath.Base(event.Name)
+					logrus.Debugf("container exited: %v", containerID)
+					c := s.GetContainer(containerID)
+					if c != nil {
+						err := s.Runtime().UpdateStatus(c)
+						if err != nil {
+							logrus.Warnf("Failed to update container status %s: %v", c, err)
```
Comment (TomSweeneyRedHat) [inline server/server.go]: nit to consider.  

logrus.Warnf("Failed to update status for container: %s, error: %v",  c, err)

**Comment by rhatdan** [comment]
@runcom PTAL

**Comment by runcom** [comment]
LGTM

**Pull request merged**